# Learning Python

I'm aspiring to become an bioinformatician. This is because i've found the act of coding, and thinking how a program should fit together scratches an itch to create and design for me. Which is a need that i can't fulfill when doing wet lab work or writing lab journal reports. 

Bioinformatics as a field is quite diverse in a way that what one bioinformatician does, could be different from what an other bioinformatician does. There are lots of different software tools you need to learn depending on which subject within biology you are working with. However i have found that most bioinformaticians use a combination of R and Python for most of their work. 



I've made a good start on R during my minor, but i dont have any experience with Python.
To start to learn Python i'm planning on going through the book [Think Python](https://greenteapress.com/wp/think-python/) by Allen B. Downey. I'ts very important to gain experience when learning a new programming language, and Think Python has exercises to play with. However these are not focused on bioinformatics. To gain experience with python for bioinformatics i want to go over [Rosalind](http://rosalind.info/about/) exercises. Rosalind is a website that has various coding problems that you could encounter as an bioinformatician, so it's a great way to apply what i've learned in python to bioinformatic problems.  

A Internship that i've applied for also recommended me three [Coursera](https://www.coursera.org/) courses that i am planning to delve into:

* [Python for Everybody Specialization](https://www.coursera.org/specializations/python)
* [An Introduction to Interactive Programming in Python (Part 1)](https://www.coursera.org/learn/interactive-python-1)
* [An Introduction to Interactive Programming in Python (Part 2)](https://www.coursera.org/learn/interactive-python-2)

## Rosalind exercises:

***

### Q1 Counting DNA nucleotides

```{python}
def count(s, c):
    count = 0
    
    for i in s:
        if c == i:
            count += 1
    return count
```

```{python}
with open("data/rosalind_dna.txt", 'r')as seq:
    seq_1 = seq.readline()
    out = []
    
    for i in ['A', 'C', 'G', 'T']:
        out.append(count(seq_1, i))
    
    print(out)
```

***

### Q2 Transcribing DNA into RNA

```{python}
with open('data/rosalind_rna.txt', 'r')as rna:
    s = rna.readline().rstrip()

u = []

for i in list(s):
    if i == 'T':
        u.append('U')
    else:
        u.append(i)
    
''.join(u)
```

***

### Q3 Complementing a Strand of DNA

```{python}
with open("data/rosalind_revc.txt", 'r') as dna:
    
    seq = dna.readline().rstrip()

com = seq.translate(''.maketrans('ATCG', 'TAGC'))
#reverse string
com[::-1]
```

***

### Q4 : Rabbits and Recurrence Relations

```{python}
def fibonacci(n, k):
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fibonacci(n-1, k) + fibonacci(n-2, k)*k 
    
```

```{python}
fibonacci(31,3)
```

***

### Q5: Computing GC Content

```{python}
#convert multi to single line
def to_sigle_fastq(in_file, out_file, header):
    """"""
    
    with open(in_file, 'r')as fastq_multi, open(out_file, 'w') as fastq_single:
        
        block = []
        
        for line in fastq_multi:
            if line.startswith(header):
                #if block contains anything, write block to file
                #this only writes the seq after a seq has been collected in block
                if block:
                    fastq_single.write(''.join(block) + '\n')
                    block = []
                #always write header line to file
                fastq_single.write(line)
            else:
                #add seq on other line to block
                block.append(line[:-1])
                
        fastq_single.write(''.join(block))
```

```{python}
to_sigle_fastq('data/rosalind_gc.txt', 'data/single.txt', '>Rosalind')
```

```{python}
#put header and seq in seprate lists with equal length
with open('data/single.txt', 'r')as fastq:
    
    id_seq = []
    seq = []
    
    for line in fastq:
        if '>Rosalind' in line:
            id_seq.append(line[:-1])
        else:
            seq.append(line[:-1])

print(id_seq, seq)

#calculating cg content
cg = []
for i in seq:
    
    cg.append((i.count('C') + i.count('G')) / len(i) * 100)
    
```

```{python}
print(cg.index(max(cg)))
print(id_seq[4])
print(cg[4])
```

***

### Q6: Counting point mutations

```{python}
def ham_distance(string1, string2):
    count = 0
    for c1, c2 in zip(string1, string2):
        if c1 != c2:
            count += 1
            
    return count
```

```{python}
with open('data/rosalind_hamm.txt', 'r')as f:
    
    t = []
    
    for line in f:
        t.append(line[:-1])

ham_distance(t[0], t[1])
```

***

### Q7 Mendels first law

```{python}
def kmn(k, m, n):
    
    total = k+m+n
    
    prob = []

    #possible events for k
    prob.append(k/total * ((k-1)/(total-1))) 
    prob.append(k/total * ((m)/(total-1)))
    prob.append(k/total * ((n)/(total-1)))

    #possible events for m
    prob.append(m/total * ((k)/(total-1))) 
    prob.append(m/total * ((m-1)/(total-1)))
    prob.append(m/total * ((n)/(total-1)))
    
    #possible events for n
    prob.append(n/total * ((k)/(total-1))) 
    prob.append(n/total * ((m)/(total-1)))
    prob.append(n/total * ((n-1)/(total-1)))

    
    #m * m
    prob[4] = prob[4]*(3/4)
    #m * n 
    prob[5] = prob[5]*(2/4)
    #n * m
    prob[7] = prob[7]*(2/4)
    #n * n
    prob[8] = prob[8]*0

    return sum(prob)
```

```{python}
kmn(28, 20, 24)
```

***

### Q8: Translating RNA into Protein

```{python}
def trans_protein(string):

    #make translate table
    table = {"UUU":"F", "UUC":"F", "UUA":"L", "UUG":"L",
    "UCU":"S", "UCC":"S", "UCA":"S", "UCG":"S",
    "UAU":"Y", "UAC":"Y", "UGU":"C", "UGC":"C", "UGG":"W",
    "CUU":"L", "CUC":"L", "CUA":"L", "CUG":"L",
    "CCU":"P", "CCC":"P", "CCA":"P", "CCG":"P",
    "CAU":"H", "CAC":"H", "CAA":"Q", "CAG":"Q",
    "CGU":"R", "CGC":"R", "CGA":"R", "CGG":"R",
    "AUU":"I", "AUC":"I", "AUA":"I", "AUG":"M",
    "ACU":"T", "ACC":"T", "ACA":"T", "ACG":"T",
    "AAU":"N", "AAC":"N", "AAA":"K", "AAG":"K",
    "AGU":"S", "AGC":"S", "AGA":"R", "AGG":"R",
    "GUU":"V", "GUC":"V", "GUA":"V", "GUG":"V",
    "GCU":"A", "GCC":"A", "GCA":"A", "GCG":"A",
    "GAU":"D", "GAC":"D", "GAA":"E", "GAG":"E",
    "GGU":"G", "GGC":"G", "GGA":"G", "GGG":"G",}
    
    # Find first instance of AUG (start codon)
    start = string.find('AUG')
    
    #split string into a list
    string_lst = [string[i:i+3] for i in range(0, len(string), 3)]
    
    #find the first instance of stop codon
    end = []
    for i in ['UAG', 'UAA', 'UGA']:
        if i in string_lst:
            end.append(string_lst.index(i))
        else:
            end.append(0)
      
    #translate strings to aminoacids
    string_lst = string_lst[0:(max(end))]
    amino_seq = ''
    for i in string_lst:
        amino_seq = amino_seq + table[i] 
    
    return amino_seq
     
```

```{python}
with open('data/rosalind_prot.txt', 'r')as file:
    line = file.readline()
    print(trans_protein(line))
```

***

### Q9: Finding a Motif in DNA

```{python}
with open('data/rosalind_subs.txt', 'r') as dna:
    
    string = dna.readline().rstrip()
    motif = dna.readline().rstrip()
    
index_start = []
y = 0
x = 0

while x != -1:
    x = string.find(motif, y)
    if x != -1:
        index_start.append(string.find(motif, y))
        y = string.find(motif, y) + 1

#adding 1 to all list elements
for i in range(len(index_start)):
    index_start[i] = index_start[i] + 1

print(*index_start, sep = ' ')
```

***

### Q10: Consensus and Profile

```{python}
def consensus(seq):
    """Creates a consensus sequence out of a list of sequences.
    
    input = List of strings
    
    output = consensus sequence and nucleotide count of input sequences
    
    example:
    ------------------
    ATGCAACT
    A: 5 1 0 0 5 5 0 0 
    C: 0 0 1 4 2 0 6 1 
    G: 1 1 6 3 0 1 0 0 
    T: 1 5 0 0 0 1 1 6"""
    
    #make a nested list of seq
    seq_list = seq
#     print(seq_list)
    array = []
    for i in range(len(seq_list)):
        array.append(list(seq_list[i]))
#     print(array)

    profile = ''
    for i in range(0 , len(array[1])):
        for a in array:
            #collect all nucleotides of the diffrent sequences, store in string
            profile = profile + a[i]

    profile_lst = [profile[i:i+len(array)] for i in range(0, len(profile), len(array))]
         
#     print(profile)
#     print(profile_lst)
    
    #count A T C G, store counts in lists
    A = []
    C = []
    G = []
    T = []
    for i in profile_lst:
        A.append(i.count('A'))
        C.append(i.count('C'))
        G.append(i.count('G'))
        T.append(i.count('T'))
    
    #create consessus
    profile = [A, C, G , T]
    chunk = []
    consens = ''
    
    for i in range(len(A)):
        for n in range(4):
            chunk.append(profile[n][i])
        consens = consens + str(chunk.index(max(chunk)))
        chunk = []
        
    #translate interger into dna string:
    table = {ord('0'):'A', ord('1'):'C', ord('2'):'G', ord('3'):'T'}
    consens = consens.translate(table)
    
    #format output
    print(consens)
    print('A:', ' '.join(str(i) for i in A), '\n' +
          'C:', ' '.join(str(i) for i in C), '\n' +
          'G:', ' '.join(str(i) for i in G), '\n' +
          'T:', ' '.join(str(i) for i in T))
    
```

A test with an example dataset:

```{python}
consensus(['ATCCAGCT', 'GGGCAACT', 'ATGGATCT', 'AAGCAACC', 'TTGGAACT', 'ATGCCATT', 'ATGGCACT'])
```

```{python}
def to_sigle_fastq(in_file, out_file, header):
    """"""
    
    with open(in_file, 'r')as fastq_multi, open(out_file, 'w') as fastq_single:
        
        block = []
        
        for line in fastq_multi:
            if line.startswith(header):
                #if block contains anything, write block to file
                #this only writes the seq after a seq has been collected in block
                if block:
                    fastq_single.write(''.join(block) + '\n')
                    block = []
                #always write header line to file
                fastq_single.write(line)
            else:
                #add seq on other line to block
                block.append(line[:-1])
                
        fastq_single.write(''.join(block)+ '\n')
```

```{python}
to_sigle_fastq('data/rosalind_cons.txt', 'data/rosalind_cons_single.txt', '>Rosalind')
```

```{python}
with open('data/rosalind_cons_single.txt', 'r')as file:
    
    string = []
    
    for line in file:
        if line[0].startswith('>'):
            pass
        else:
            string.append(line[:-1])
            
    consensus(string)
```





